src/libstrongswan/plugins/openssl/openssl_kem.c

```diff
@@ -0,0 +1,371 @@
+/*
+ * Copyright (C) 2024 Gerardo Ravago
+ *
+ * Copyright (C) secunet Security Networks AG
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.  See <http://www.fsf.org/copyleft/gpl.txt>.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * for more details.
+ */
+
+
+#include "openssl_kem.h"
+
+#include <openssl/evp.h>
+
+#ifdef OPENSSL_IS_AWSLC
+#include <openssl/experimental/kem_deterministic_api.h>
+
+typedef struct private_key_exchange_t private_key_exchange_t;
+
+/**
+ * Private data of an key_exchange_t object.
+ */
+struct private_key_exchange_t {
+	/**
+	 * Public interface.
+	 */
+	key_exchange_t public;
+
+	/**
+	 * Key Exchange Method Transform ID.
+	 */
+	key_exchange_method_t group;
+
+	/**
+	 * OpenSSL EVP_PKEY object for a KEM Keypair. Only set on initiator.
+	 */
+	EVP_PKEY *pkey;
+
+	/**
+	 * Shared secret
+	 */
+	chunk_t shared_secret;
+
+	/**
+	 * Ciphertext. Used as the "public key" for the responder.
+	 */
+	chunk_t ciphertext;
+
+	/**
+	 * Deterministic Random Bit Generator (DRBG)
+	 */
+	drbg_t *drbg;
+};
+
+/**
+ * Return the OpenSSL NID for to this instance's key_exchange_method_t.
+ */
+static int openssl_kem_get_nid(private_key_exchange_t *this)
+{
+	switch (this->group) {
+		case ML_KEM_512:
+			return NID_MLKEM512;
+		case ML_KEM_768:
+			return NID_MLKEM768;
+		case ML_KEM_1024:
+			return NID_MLKEM1024;
+		default:
+			return NID_undef;
+	}
+}
+
+METHOD(key_exchange_t, get_method, key_exchange_method_t,
+	private_key_exchange_t *this)
+{
+	return this->group;
+}
+
+/**
+ * Call OpenSSL's KEM KeyGen method and store a reference to the key.
+ */
+static bool openssl_kem_generate_pkey(private_key_exchange_t *this)
+{
+	EVP_PKEY_CTX *ctx = NULL;
+	size_t seed_length = 0;
+	chunk_t seed = chunk_empty;
+
+	ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_KEM, NULL);
+	if (!ctx)
+	{
+		return FALSE;
+	}
+	if (!EVP_PKEY_CTX_kem_set_params(ctx, openssl_kem_get_nid(this)))
+	{
+		EVP_PKEY_CTX_free(ctx);
+		return FALSE;
+	}
+	if (!EVP_PKEY_keygen_init(ctx))
+	{
+		EVP_PKEY_CTX_free(ctx);
+		return FALSE;
+	}
+	if (this->drbg)
+	{
+		if (!EVP_PKEY_keygen_deterministic(ctx, NULL, NULL, &seed_length))
+		{
+			EVP_PKEY_CTX_free(ctx);
+			return FALSE;
+		}
+		seed = chunk_alloc(seed_length);
+		if (!this->drbg->generate(this->drbg, seed.len, seed.ptr))
+		{
+			EVP_PKEY_CTX_free(ctx);
+			return FALSE;
+		}
+		if (!EVP_PKEY_keygen_deterministic(ctx, &this->pkey, seed.ptr,
+										   &seed.len))
+		{
+			EVP_PKEY_CTX_free(ctx);
+			EVP_PKEY_free(this->pkey);
```

Comment:
Please remove these remaining `EVP_PKEY_free(this->pkey)` calls. Should the object be allocated, it's freed in `destroy()` anyway, so there is no need to do it here. And as I mentioned before, calling it here as well will cause a double-free because `EVP_PKEY_free()` won't set the pointer to NULL